### ISSUE TITLE
v2.7.0 CLI argo logs for workflows are broken

### ISSUE THREAD

**Issue by samath117** [argoproj/argo#2589]
Checklist:

 * [x] I've included the version.
 * [x] I've included reproduction steps.
 * [x] I've included the workflow YAML.
 * [x] I've included the logs.
 
**What happened**:

`argo logs` in v2.7.0 for workflows are broken in many ways.

1. Some pods' logs are dropped, seemingly at random.
2. The logs from each pod are arranged in random order, not chronological.
3. Run with the `-f` flag, `argo logs` misses logs from pods that are started after the command is issued.
4. The `-w` flag is deprecated but also rejected, another breaking change.
5. The logs are labeled with the pod names, making the color scheme useless (they all get the same color).

**How to reproduce it (as minimally and precisely as possible)**:

Just run a simple script with multiple steps with logs.

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: logs-problems-
spec:
  entrypoint: plan
  templates:
  - name: plan
    steps:

    - - name: report-1
        template: report
        arguments:
          parameters:
          - name: message
            value: one

    - - name: report-2
        template: report
        arguments:
          parameters:
          - name: message
            value: two

    - - name: report-3
        template: report
        arguments:
          parameters:
          - name: message
            value: three

    - - name: report-4
        template: report
        arguments:
          parameters:
          - name: message
            value: four

    - - name: report-5
        template: report
        arguments:
          parameters:
          - name: message
            value: five

  - name: report
    inputs:
      parameters:
      - name: message
    container:
      image: alpine:3.11
      command: [sh, -c]
      args: ["echo {{inputs.parameters.message}}"]
```

Here's literally the output I see on the command line:
```
$ argo logs -n virtual logs-problems-jfcrs -f
logs-problems-jfcrs-1210664: two
logs-problems-jfcrs-1959331710: three
logs-problems-jfcrs-1789405150: one
$ argo logs -n virtual logs-problems-jfcrs
logs-problems-jfcrs-1210664: two
logs-problems-jfcrs-1959331710: three
$ argo logs -n virtual logs-problems-jfcrs
logs-problems-jfcrs-1959331710: three
logs-problems-jfcrs-1043726436: four
logs-problems-jfcrs-1210664: two
logs-problems-jfcrs-3944066254: five
$ argo logs -n virtual logs-problems-jfcrs
logs-problems-jfcrs-3944066254: five
logs-problems-jfcrs-1789405150: one
logs-problems-jfcrs-1210664: two
logs-problems-jfcrs-1043726436: four
```

You can see that I ran the `-f` command before the fourth and fifth pods had started, so it ignores those. (I've been able to reproduce this pattern over multiple attempts.) The later calls without that flag grab a random subset of the logs and paste them in a random order. Moreover, all of those logs were printed in red, a useless color scheme.

Needless to say, this makes `argo logs` on the CLI completely unusable in v2.7.0.

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by simster7** [comment]
@alexec Could you take a look at this?

**Comment by alexec** [comment]
Monday. In the meantime you could use ‘kubectl logs’. Thank you for a clear set of issues and the YAML to repro.

**Comment by alexec** [comment]
@samath117 is this a regression from previous behaviour do you think?

**Comment by alexec** [comment]
**Some pods' logs are dropped, seemingly at random. **

We either (a) need me information to diagnose this or (b) time-box analysis.

**The logs from each pod are arranged in random order, not chronological.** 

Do you mean the log entries themselves? Or just the logs.

**Run with the -f flag, argo logs misses logs from pods that are started after the command is issued.**

This is a bug.

**The -w flag is deprecated but also rejected, another breaking change.**

I'm not sure what to do about this. This is a breaking change, but sometimes it is better not to fix a breaking change. 

In this case, if you want to get the logs of a single pod, you can use `kubectl logs PODNAME`. This will always be a better way to do this task. There is an edge-case where you might not have been able to do this, when you running v2.5+ with Argo Server and you removed RBAC for pod logs from the user. I don't see this as likely.

That's not to say we should not fix this breaking change, just some fuel for discussion.

**The logs are labeled with the pod names, making the color scheme useless (they all get the same color).**

This is a bug is logs.go#L103. Easy fix.

**Comment by alexec** [comment]
There is a bug where if a pod in the workflow finishes, it stop listening for new pods 🤦‍♂

**Comment by samath117** [comment]
> is this a regression from previous behaviour do you think?

Yes. Here is what I see for the logs when I run this in Argo v2.6.3:
```
report-1:       one
report-2:       two
report-3:       three
report-4:       four
report-5:       five
```
logs-problems-jfcrs-1789405150: one
logs-problems-jfcrs-1210664: two
logs-problems-jfcrs-1959331710: three
logs-problems-jfcrs-1043726436: four
logs-problems-jfcrs-3944066254: five
(But even more colorful!)

Note also that they are annotated based on the lowest-level step name, not the pod name. Personally, I find the step names more readable, but the pod names are less ambiguous. Could we have both? Something like:
```
logs-problems-jfcrs-1789405150 (report-1): one
logs-problems-jfcrs-1210664 (report-2): two
logs-problems-jfcrs-1959331710 (report-3): three
logs-problems-jfcrs-1043726436 (report-4): four
logs-problems-jfcrs-3944066254 (report-5): five
```

> > The logs from each pod are arranged in random order, not chronological.

> Do you mean the log entries themselves? Or just the logs.

Both. In previous versions, if the workflow had multiple pods that were running simultaneously, the workflow logs would interleave each of the log entries based on when they came in. This was very helpful for debugging complicated workflows. In this example, there is just one entry for each pod and they still show up in a random order, but I would also like to retain the previous functionality of interspersed log entries, all ordered by timestamp. I'll come up with a new example in another comment to make this clear.

> > The -w flag is deprecated but also rejected, another breaking change.

> I'm not sure what to do about this. This is a breaking change, but sometimes it is better not to fix a breaking change.

Can you just make the `-w` flag have no effect? That way any old code using `-w` to get workflow logs will still work.

> In this case, if you want to get the logs of a single pod, you can use kubectl logs PODNAME. This will always be a better way to do this task.

Again, this doesn't get all logs from a workflow, or order them by timestamp.

**Comment by samath117** [comment]
Here is a more complicated example to make sure the timing and interweaving looks right:
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: logs-problems-
spec:
  entrypoint: plan
  templates:
  - name: plan
    steps:

    - - name: report-1
        template: report
        arguments:
          parameters:
          - name: before-delay
            value: 0
          - name: message
            value: one

      - name: report-2
        template: report
        arguments:
          parameters:
          - name: before-delay
            value: 10
          - name: message
            value: two

      - name: report-3
        template: report
        arguments:
          parameters:
          - name: before-delay
            value: 20
          - name: message
            value: three

      - name: report-4
        template: report
        arguments:
          parameters:
          - name: before-delay
            value: 30
          - name: message
            value: four

      - name: report-5
        template: report
        arguments:
          parameters:
          - name: before-delay
            value: 40
          - name: message
            value: five

  - name: report
    inputs:
      parameters:
      - name: message
      - name: before-delay
      - name: between-delay
        value: 15
    container:
      image: alpine:3.11
      command: [sh, -c]
      args: ["sleep {{inputs.parameters.before-delay}}s; \
              echo {{inputs.parameters.message}}; \
              sleep {{inputs.parameters.between-delay}}s; \
              echo {{inputs.parameters.message}}-again"]
```

Running this in v2.6.3, here are the workflow logs:
```
report-1:       one
report-2:       two
report-1:       one-again
report-3:       three
report-2:       two-again
report-4:       four
report-3:       three-again
report-5:       five
report-4:       four-again
report-5:       five-again
```
Notice how the numbers are nicely interwoven exactly as you would expect based on those `sleep` commands. (Total delays are, respectively, 0, 10, 15, 20, 25, 30, 35, 40, 45, 55.)

In Argo v2.7.0, here's what I get (showing a few attempts)
```
logs-problems-6q9v9-2155440197: five
logs-problems-6q9v9-2155440197: five-again
logs-problems-6q9v9-2138662578: four
logs-problems-6q9v9-2138662578: four-again
logs-problems-6q9v9-2088329721: one
```
```
logs-problems-6q9v9-2155440197: five
logs-problems-6q9v9-2155440197: five-again
logs-problems-6q9v9-2088329721: one
logs-problems-6q9v9-2088329721: one-again
```
```
logs-problems-6q9v9-2054774483: three
logs-problems-6q9v9-2054774483: three-again
logs-problems-6q9v9-2037996864: two
logs-problems-6q9v9-2037996864: two-again
logs-problems-6q9v9-2138662578: four
logs-problems-6q9v9-2138662578: four-again
logs-problems-6q9v9-2155440197: five
```
Among the dropped lines, I do notice that (as you can see in these examples) the `again` part is only dropped if it's the last pod listed, which suggests that something is prematurely exiting.

Beyond the dropping issue, I would really like to see the same sorting of log lines by timestamp that v2.6.3 and prior versions had. It seems that the new logs makes no attempt to do this, and just reports all of the pods in a jumbled-up order.

**Comment by alexec** [comment]
@samath117 there is a a linked PR for this, that should fix some of the issues. There is an ordering issue that affects pods that are running when the logging starts that is related to Go routine scheduling. I'll need to investigate further.

**Comment by alexec** [comment]
All items are now fixed with the exception of the `-w` flag which I personally don't think we should fix. I'm open to a good argument to fix it - one beyond it is a breaking change.

**Comment by simster7** [comment]
Closing as fixed in #2595 and #2639
